pkg/app/launcher/cmd/launcher/launcher.go

```diff
@@ -511,15 +514,72 @@ func makePipedArgs(launcherArgs []string, configFile string) []string {
 	return pipedArgs
 }
 
-func parseConfig(data []byte) (*config.PipedSpec, error) {
-	cfg, err := config.DecodeYAML(data)
+func parseConfig(data []byte) (*LauncherSpec, error) {
+	js, err := yaml.YAMLToJSON(data)
 	if err != nil {
 		return nil, err
 	}
-	if cfg.Kind != config.KindPiped {
-		return nil, fmt.Errorf("wrong configuration kind for piped: %v", cfg.Kind)
+
+	c := &Config{}
+	if err := json.Unmarshal(js, c); err != nil {
+		return nil, err
+	}
+
+	if err := c.Validate(); err != nil {
+		return nil, err
+	}
+	return &c.Spec, nil
+}
+
+type Config struct {
+	Kind       config.Kind  `json:"kind"`
+	APIVersion string       `json:"apiVersion,omitempty"`
+	Spec       LauncherSpec `json:"spec"`
+}
+
+func (c *Config) Validate() error {
+	if c.Kind != config.KindPiped {
+		return fmt.Errorf("wrong configuration kind for piped: %v", c.Kind)
+	}
+	if c.Spec.ProjectID == "" {
+		return errors.New("projectID must be set")
+	}
+	if c.Spec.PipedID == "" {
+		return errors.New("pipedID must be set")
+	}
+	if c.Spec.PipedKeyData == "" && c.Spec.PipedKeyFile == "" {
+		return errors.New("either pipedKeyFile or pipedKeyData must be set")
+	}
+	if c.Spec.PipedKeyData != "" && c.Spec.PipedKeyFile != "" {
+		return errors.New("only pipedKeyFile or pipedKeyData can be set")
+	}
+	if c.Spec.APIAddress == "" {
+		return errors.New("apiAddress must be set")
+	}
+	return nil
+}
+
+type LauncherSpec struct {
```

Comment:
nits, is it enough to make this struct just a private struct?

Reply:
I thought it would be okay to define it in config. but it's up to you.

Reply:
Sure. Fixed them. 👍

Reply:
> I thought it would be okay to define it in /config. but it's up to you.

I feel you. That is nice too. Let me apply your suggestion. 👍